pkg/rpm/database.go

```diff
@@ -0,0 +1,285 @@
+	rpmStderr      *bytes.Buffer
+	rpmStdout      io.ReadCloser
+	// Query state.
+	err         error
+	nextPackage rpmPackage
+}
+
+// rpmPackage represents an RPM package information.
+type rpmPackage struct {
+	Name      string
+	Version   string
+	Arch      string
+	Module    string
+	Filenames []string
+}
+
+// QueryOpts is a set of options for database queries.
+type QueryOpts struct {
+	// Testing if true, performs a test query, which currently uses a different query
+	// format for compatibility with old rpm versions.
+	Testing bool
+}
+
+func init() {
+	if features.RHEL9Scanning.Enabled() {
+		// For RHEL9 we need to enable sqlite database (rpm >= 4.16)
+		databaseFiles.Add("rpmdb.sqlite")
+	}
+}
+
+// DatabaseFiles returns a slice containing full paths of all RPM database
+// files known for the different backend we support. The paths are relative to
+// root.
+func DatabaseFiles() []string {
+	var paths []string
+	for _, filename := range databaseFiles.AsSlice() {
+		paths = append(paths, path.Join(databaseDir, filename))
+	}
+	return paths
+}
+
+// CreateDatabaseFromLayer creates an RPM database in a temporary directory
+// from the RPM database found in the container image. All known RPM database
+// backend is supported (i.e. bdb, sqlite). If no database is found in the image,
+// returns nil.
+func CreateDatabaseFromLayer(imageFiles tarutil.LayerFiles) (*rpmDatabase, error) {
+	// Find all known RPM database models and their files. It is unlikely that the
+	// image will contain more than one model, but in that scenario we copy all files
+	// and rely on the fact that `rpm` will select the most up-to-date database
+	// model, instead of replicating that knowledge in the code.
+	dbFiles := make(map[string]tarutil.FileData)
+	for _, name := range databaseFiles.AsSlice() {
+		if data, exists := imageFiles.Get(path.Join(databaseDir, name)); exists {
+			dbFiles[name] = data
+		}
+	}
+	if len(dbFiles) == 0 {
+		// Not rpm database was found.
+		return nil, nil
+	}
+	// Write the database files to the filesystem.
+	dbDir, err := os.MkdirTemp("", "rpm")
+	if err != nil {
+		logrus.WithError(err).Error("could not create temporary folder for the rpm database")
+		return nil, commonerr.ErrFilesystem
+	}
+	defer func() {
+		// Remove temporary directory if returning on errors.
+		if err != nil {
+			_ = os.RemoveAll(dbDir)
+		}
+	}()
+	for name, data := range dbFiles {
+		dbFilename := filepath.Join(dbDir, name)
+		err = os.WriteFile(dbFilename, data.Contents, 0700)
+		if err != nil {
+			logrus.WithError(err).Error("failed to create rpm database file")
+			return nil, commonerr.ErrFilesystem
+		}
+	}
+	return &rpmDatabase{
+		dbPath: dbDir,
+	}, nil
+}
+
+// QueryAll starts a query for all packages in the RPM database, returns a query
+// iterator. A query is actually performed by an underlying rpm sub-process, and
+// return is retrieved after parsing its output and status.
+func (d *rpmDatabase) QueryAll(opts QueryOpts) *rpmDatabaseQuery {
```

Comment:
If we want to mimic something similar to https://pkg.go.dev/database/sql#DB.Query, then let's return `(*rpmDatabaseQuery, error)`

Reply:
This will force the code to have one additional `if err != nil {` unnecessarily. Why do we need to follow `DB.Query` strictly even when it doesn't make sense to us?